lib/synapse/service_watcher/zookeeper/zookeeper.rb

```diff
@@ -54,29 +55,33 @@ def initialize(opts={}, synapse, reconfigure_callback)
       @zk_hosts = zk_host_list.join(',')
 
       @zk = nil
+      @watcher = nil
+      @thread = nil
+      @should_exit = Synapse::AtomicValue.new(false)
     end
 
     def start
-      @watcher = nil
-
       log.info "synapse: starting ZK watcher #{@name} @ cluster: #{@zk_cluster} path: #{@discovery['path']} retry policy: #{@retry_policy}"
+
+      # Zookeeper processing is run in a background thread so that any retries
+      # do not block the main thread.
       zk_connect do
-        # the path must exist, otherwise watch callbacks will not work
-        with_retry(@retry_policy.merge({'retriable_errors' => ZK_RETRIABLE_ERRORS})) do |attempts|
-          unless @zk.exists?(@discovery['path'])
-            statsd_time('synapse.watcher.zk.create_path.elapsed_time', ["zk_cluster:#{@zk_cluster}", "service_name:#{@name}"]) do
-              log.info "synapse: zk create at #{@discovery['path']} for #{attempts} times"
-              create(@discovery['path'])
-            end
-          end
-        end
+        @thread = Thread.new {
+          start_discovery
```

Comment:
if the background thread is performing retry but there is a signal of `@should_exit`, it won't know until finishing retry, right? do you think it is better to move `@should_exit.get` into the retry block?

Reply:
This is an existing problem which I wasn’t sure how to solve cleanly. The retry code is spread over a bunch of places and abstracted away into the `with_retry` function. So it’s hard to integrate the clean exit into there.